specs/agents/sanitization.md

```diff
@@ -1,16 +1,58 @@
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to
+be interpreted as described in
+[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
 ## Data sanitization
 
 ### `sanitize_field_names` configuration
 
 Sometimes it is necessary to sanitize, i.e., remove,
 sensitive data sent to Elastic APM.
-This config accepts a list of wildcard patterns of field names which should be sanitized.
-These apply to HTTP headers (including cookies) and `application/x-www-form-urlencoded` data (POST form fields).
-The query string and the captured request body (such as `application/json` data) will not get sanitized.
+
+This config accepts a list of wildcard patterns of field names which control
+how an agent will sanitize data.
 
 |                |   |
 |----------------|---|
 | Type           | `List<`[`WildcardMatcher`](../../tests/agents/json-specs/wildcard_matcher_tests.json)`>` |
 | Default        | `password, passwd, pwd, secret, *key, *token*, *session*, *credit*, *card*, authorization, set-cookie` |
 | Dynamic        | `true` |
 | Central config | `true` |
+
+## Configuration
+
+Agents MUST provide a minimum default configuration of
+
+    [ 'password', 'passwd', 'pwd', 'secret', '*key', '*token*', '*session*',
+      '*credit*','*card*', 'authorization', 'set-cookie']
+
+for the `sanitize_field_names` configuration value.  Agent's MAY include the
+following extra fields in their default configuration to avoid breaking changes
+
+    ['pw','pass','connect.sid']
+
+If an end-user configures different values, these values MUST **replace** the
+above values. An end-user's configured values MUST NOT be merged with these
+default values.
+
+## Sanitizing Values
+
+If a payload field's name (a header key, a form key) matches a configured
+wildcard, that field's _value_ MUST be removed/redacted and the key itself
+MUST still be reported in the agent payload. Agents MAY chose the string
+they use to replace the value so long as it's consistent and does not reveal
+the value it has replaced. Some example replacement strings
+include `REDACTED`, `**********`, `-----------`, etc.
+
+Fields that MUST be sanitized are the HTTP Request headers, HTTP Response
+headers, and form fields in an `application/x-www-form-urlencoded` request
+body.  No fields (including `set-cookie` headers) are exempt from this.
+
+The query string and other captured request bodies (such as `application/json`)
+MUST NOT be sanitized.
```

Comment:
There may be agents that do that (not Java). 
Why not allowing this (assuming overhead is not a concern)? If there was already a discussion where it was agreed to ban this, good enough for me.

Reply:
There was some earlier discussion [about query strings](https://github.com/elastic/apm/pull/334#issuecomment-688212621), and since the initial spec singled out `application/x-www-form-urlencoded` it also seemed to imply that other requests body were not to be sanitized.  If it turns out other agents _do_ currently sanitize these values we can shift this from a MUST NOT to a MAY.

Reply:
We can decide based on what agents do, but what I am suggesting is to allow this regardless.

Reply:
Overhead is the main concern here. Redacting a specific key in the JSON request body should be done on the APM Server or in and Elasticsearch ingest node.

Reply:
I assumed this is the main concern, however it can still be allowed for cases where it is not a concern or it is a requirement (e.g. a customer for whom redacting in APM Server/ES is too late).
But if no agent needs to spend effort on removing something because of this limitation, not harm starting this way.

Reply:
Thanks for @eyalkoren and @felixbarny for the feedback here.  Based on what you've both said I've changed this to read 

> The query string and other captured request bodies (such as `application/json`)
SHOULD NOT be sanitized.

Stated like this it discourages sanitizing these values (out of an concern for overhead/perf.), but allows agent team who have weighed the trade-offs and decided that it's worth doing  to keep their implementations. 

If there's more to discuss here feel free to unresolve the conversation.